sdk/go/pulumix/compose.go

```diff
@@ -0,0 +1,198 @@
+//     If it does, it MUST NOT use the Composer in those goroutines.
+func Compose[T any](ctx context.Context, f func(*Composer) (T, error)) Output[T] {
+	var wg internal.WorkGroup
+	outputState := internal.NewOutputState(&wg, typeOf[T]())
+	c := Composer{
+		ctx:         ctx,
+		known:       true,
+		secret:      false,
+		outputState: outputState,
+	}
+
+	go func() {
+		defer func() {
+			// If f kills this goroutine before returning,
+			// it was because of one of two reasons:
+			//
+			//  - ComposeAwait was called with an unknown or failed input
+			//    which killed the goroutine but fulfilled the output state.
+			//  - The user killed the goroutine with a panic
+			//    or by calling runtime.Goexit().
+			//
+			// For the latter case, to avoid a deadlock
+			// we must fulfill the output state before exiting.
+			if !c.fulfilled {
+				var err error
+				if pval := recover(); pval != nil {
+					err = fmt.Errorf("panic: %v\n%s", pval, debug.Stack())
+				} else {
+					err = errors.New("goroutine exited before returning")
+				}
+				internal.RejectOutput(outputState, err)
+			}
+
+			// After the function returns, zero out the Composer
+			// to protect against misuse like storing a pointer
+			// to the Composer outside f.
+			// e.g.,
+			//
+			//	var c Composer
+			//	o := Compose(ctx, func(c2 *Composer) (O, error) {
+			//		c = *c2
+			//		// ...
+			//	})
+			c = Composer{}
+		}()
+
+		v, err := f(&c)
+		if err != nil {
+			internal.RejectOutput(outputState, err)
+		} else {
+			internal.FulfillOutput(outputState, v, c.known, c.secret, c.deps, nil)
+		}
+	}()
+
+	return Output[T]{OutputState: outputState}
+}
+
+// ComposeAwait awaits for the output of the given input and returns it.
+//
+//	var o pulumix.Output[T] = someOutput
+//	v := pulumix.ComposeAwait(c, o)
+//	// v is of type T
+//
+// Use this to combine multiple outputs into a single value.
+//
+//	i, err := strconv.ParseInt(pulumix.ComposeAwait(c, strOutput))
+//	if err != nil {
+//		return 0, err
+//	}
+//
+// If the input is unknown or failed,
+// ComposeAwait will cancel the entire composition operation.
+// For example, given the following,
+//
+//	v1 := pulumix.ComposeAwait(c, o1)
+//	v2 := pulumix.ComposeAwait(c, o2)
+//	return f(v1, v2)
+//
+// If either o1 or o2 is unknown or failed,
+// the composition will abort and f will not be called.
+//
+// In using ComposeAwait, be aware that:
+//
+//   - It can only be called inside a [Compose] call.
+//   - It MUST NOT be called from goroutines spawned by f.
+func ComposeAwait[T any](c *Composer, o Input[T]) T {
+	contract.Assertf(c.outputState != nil, "ComposeAwait called outside Compose")
+
+	v, known, secret, deps, err := await(c.ctx, o.ToOutput(c.ctx))
+	c.secret = c.secret || secret
+	c.known = c.known && known
+	c.deps = append(c.deps, deps...)
+	if err != nil || !known {
+		var zero T
+		internal.FulfillOutput(c.outputState, zero, false, c.secret, c.deps, err)
+		c.fulfilled = true
+		runtime.Goexit()
```

Comment:
Oof interesting so runtime.Goexit() terminates the flow, and that was captured with `recover` elsewhere in this PR. I guess this is the only way given the signature T. Is there  a way to distinguish ComposeAwait-initiated interrupts from general interrupts? E.g. so that this works:

```
x := ComposeAwait(c, o) // o resolves to an error, we're fine
``` 

But this panics:

```
panic("arbitrary panic")
x := ComposeAwait(c, o) // o resolves to an error, we're fine
```

Perhaps use `panic(customType{})` or some such trick?

BTW TIL one can recover from runtime.Goexit() I didn't know that.

Reply:
Oh..
```
Goexit terminates the goroutine that calls it. No other goroutine is affected. Goexit runs all deferred calls before terminating the goroutine. Because Goexit is not a panic, any recover calls in those deferred functions will return nil.

Calling Goexit from the main goroutine terminates that goroutine without func main returning. Since func main has not returned, the program continues execution of other goroutines. If all other goroutines exit, the program crashes
```

Reply:
panics and runtime.Goexit are different.
They'll both run deferred code. runtime.Goexit *will not* panic.
It'll just stop running the current goroutine.
So yeah, there is a way to differentiate between panics and not.

Reply:
Just read up on Go docs here, great to learn.